Cumulus 1560 features (#1312)
nasa/cumulus: docs/features/data_in_dynamodb.md

---

# Cumulus Metadata in DynamoDB

* [DynamoDB Backup and Restore](#backup-and-restore-with-aws)
* [DynamoDB Auto Scaling](#dynamodb-auto-scaling)

[@cumulus/api](https://www.npmjs.com/package/@cumulus/api) uses a number of methods to preserve the metadata generated in a Cumulus instance.

All configurations and system-generated metadata is stored in DynamoDB tables except the logs. System logs are stored in the AWS CloudWatch service.

> `@cumulus/api` uses the following DynamoDB tables:

Reviewer: 👍

## Enabling PITR during deployment

By default, the Cumulus [data-persistence module](https://github.com/nasa/cumulus/blob/master/tf-modules/data-persistence) enables PITR on the default tables listed in the [module's variable defaults](https://github.com/nasa/cumulus/blob/master/tf-modules/data-persistence/variables.tf) for `enable_point_in_time_tables`.  At the time of writing, that list includes:

* CollectionsTable
* ExecutionsTable
* FilesTable
* GranulesTable
* PdrsTable
* ProvidersTable
* RulesTable
* UsersTable

If you wish to change this list, simply update your deployment's `data_persistence` module ([here](https://github.com/nasa/cumulus-template-deploy/blob/master/data-persistence-tf/main.tf) in the `template-deploy` repository) to pass the correct list of tables.

## Backup and Restore with cumulus-api CLI

cumulus-api CLI also includes a backup and restore command. The CLI backup command downloads the content of any of your DynamoDB tables to `.json` files. You can also use these `.json` files to restore the records to another DynamoDB table.

### Backup with the CLI

> To backup a table with the CLI, make sure `@cumulus/api` package is installed using [npm](https://www.npmjs.com/), making sure to reference the correct version:

Reviewer: ```suggestion
To backup a table with the CLI, install the `@cumulus/api` package using [npm](https://www.npmjs.com/), making sure to install the same version as your Cumulus deployment:
```

To backup a table with the CLI, make sure `@cumulus/api` package is installed. Then run:

> ```bash

Reviewer: So people will still have to install the API package to do backup/restore, even though that package is no longer required for deploying Cumulus? Thus, they still need npm?

Author: They don't *have* to, but they *can*.

Reviewer: Since node/npm are no longer in the main deployment docs, do we at least need a basic primer here or mention them as dependencies? Otherwise how to install `@cumulus/api` might not be clear

Author: Hmm, fair enough.   How about e29b4cf3d?

Reviewer: Had a minor suggestion

Author: 👍

     ./node_modules/.bin/cumulus-api backup --table <table-name>
```

the backup will be stored at `backups/<table-name>.json`

### Restore with the CLI

To restore data from a json file run the following command:

```bash
     ./node_modules/.bin/cumulus-api restore backups/<table-name>.json --table <new-table-name>
```

## DynamoDB Auto Scaling

Cumulus deployed tables from the [data-persistence module](https://github.com/nasa/cumulus/blob/master/tf-modules/data-persistence) are set to [`on-demand'](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadWriteCapacityMode.html#HowItWorks.OnDemand) mode.

> ### Important Note

Reviewer: Is this note still relevant since we're not using autoscaling anymore on the DB tables?

Author: No, indeed not.   42dd52457
